include/CL/sycl/parallelism/detail/parallelism.hpp

```diff
@@ -239,35 +239,47 @@ void parallel_for_workitem(const group<Dimensions> &g,
 
      The issue is that the parallel_for_workitem() execution is slow even
      when nd_item::barrier() is not used
-
-     \todo Simplify by just using omp parallel for collapse
   */
-
+  T_Item index{g.get_nd_range()};
   range<Dimensions> l_r = g.get_nd_range().get_local_range();
-  auto tot = l_r.get(0);
-  for (int i = 1; i < (int) Dimensions; ++i) {
-    tot *= l_r.get(i);
-  }
-#pragma omp parallel num_threads(tot)
-  {
-    T_Item index { g.get_nd_range() };
-    id<Dimensions> local; // to initialize correctly
-#pragma omp for nowait
-    for (std::size_t th_id = 0; th_id < tot; ++th_id) {
-      if (Dimensions == 1) {
-        local[0] = th_id;
-      } else if (Dimensions == 2) {
-        local[0] = th_id / l_r.get(1);
-        local[1] = th_id % l_r.get(1);
-      } else if (Dimensions == 3) {
-        local[0] = th_id / (l_r.get(1)*l_r.get(2));
-        local[1] = (th_id / l_r.get(2)) % l_r.get(1);
-        local[2] = th_id % l_r.get(2);
-      }
-      index.set_local(local);
-      index.set_global(local + id<Dimensions>(l_r)*g.get_id());
+  id<Dimensions> temp, id_l_r = id<Dimensions>(l_r);
+
+  if constexpr (Dimensions == 1) {
+#pragma omp parallel for collapse(1) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i) {
+      temp[0] = i;
+      index.set_local(temp);
+      temp[0] = i + id_l_r[0] * g.get_id()[0];
+      index.set_global(temp);
```

Comment:
`index.set_global(i + id_l_r[0] * g.get_id()[0]);` or `index.set_global({ i + id_l_r[0] * g.get_id()[0] });` ?

Reply:
Or see at the end.